### ISSUE TITLE
abuse of DT_GNU_HASH descriptor

### ISSUE THREAD

**Issue by jreiser** [rust-lang/rust#100859]
Some executable Rust programs in ELF file format for Linux, often involving the **musl** low-level library, use undocumented "features" of the DT_GNU_HASH descriptor for run-time symbol information in an ELF file.  The DT_GNU_HASH descriptor itself is undocumented by GNU binutils, so the "corner case" usage by Rust compounds the problem of deterioration of software environment by not agreeing or communicating intended meaning, and proliferation of inconsistent private interpretations.  This GitHub Issue documents problematic occurrences that I have seen, and requests that Rust contribute to creating the documentation of DT_GNU_HASH, including explanation of "corner cases" seen so far.

The cases appeared when users of UPX (https://github.com/upx/upx) reported that UPX did not understand various Rust-compiled binary executable programs.  UPX wants to understand compiled binary executables in order to compress and de-compress them, and a program's use of specific global symbols can affect UPX; so UPX must lookup those symbols, which are expected to be encoded using DT_HASH and/or DT_GNU_HASH descriptor.  The specific DT_GNU_HASH descriptors used by Rust+musl have presented unusual encodings.

1. https://github.com/upx/upx/issues/568  DT_GNU_HASH descriptor with no symbol information `(0 == nbucket)`, in a dynamically-linked executable main program that does contain global symbols.  How should this be interpreted?  Is it an attempt to save space, and force a fall-back to linear search of DT_SYMTAB?  If so, then why not save even more space by omitting DT_GNU_HASH entirely?
2. https://github.com/upx/upx/issues/525  An earlier case of DT_GNU_HASH descriptor with no symbol information, but encoded as `(1==n_bucket  && 0==buckets[0] &&  1==n_bitmask && 0==bitmask[0])`.  Save even more space by omitting DT_GNU_HASH entirely.
3. https://github.com/upx/upx/issues/476  First-seen case of DT_GNU_HASH descriptor with no symbol information `(0==nbucket)`, in a statically-linked main program with no symbols.  Save even more space by omitting DT_GNU_HASH entirely.
4. https://github.com/upx/upx/issues/369  `hash_array[j]` is not present if `bucket[k]` is zero for all `k > j`.  So the tail of a complete `hash_array[]` (which should have `nbucket` entries, just like `buckets[]`) would overlap the following table (typically DT_SYMTAB).

The `(0==nbuckets)` cases suggest that the input to UPX might have been compromised by malware clobbering the first word of DT_GNU_HASH (and then using the remaining original space as payload for malware); else why was the trivially-shorter encoding of "no DT_GNU_HASH at all" not used?  The last case (`hash_array[]` truncated by the start of DT_SYMTAB) also casts doubt on the integrity of the file.  These optimizations for reducing space could well be blessed, _if documented and explained_.  Please lend your weight to getting DT_GNU_HASH documented and annotated.

**Comment by JMLX42** [comment]
I have a similar error:

```
$ upx target/release/gltf_live_server
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2024
UPX 4.2.4       Markus Oberhumer, Laszlo Molnar & John Reiser    May 9th 2024

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
upx: target/release/gltf_live_server: CantPackException: bad DT_GNU_HASH n_bucket=0x1  n_bitmask=0x1  len=0x20  r=6

Packed 0 files.
```

In this case: `n_bucket=0x1  n_bitmask=0x1  len=0x20  r=6`

**Comment by murlakatamenka** [comment]
> But I am **not** using musl.
> 
> Update: I was using `mold` as the linker. Otherwise, `upx` works as expected.

Can confirm.

Was building [rathole](https://github.com/rapiz1/rathole/blob/main/docs/build-guide.md#minimalize-the-binary) and `upx`-ing the final binary gave that `CantPackException: bad DT_GNU_HASH` error:

```sh
UPX 4.2.4       Markus Oberhumer, Laszlo Molnar & John Reiser    May 9th 2024

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
upx: target/release/rathole: CantPackException: bad DT_GNU_HASH n_bucket=0x1  n_bitmask=0x1  len=0x20  r=6
```

Disabling `mold` makes the binary `upx`-able.

**Comment by jreiser** [comment]
For many months there has been disagreement over how to represent an "empty" DT_GNU_HASH element.  Apparently the simplest method (completely omitting the entire DT_GNU_HASH descriptor) is not acceptable to some other softwares, and the disagreement is over the exact byte-wise representation of a "present but empty" element.

UPX has tried a few methods, and its current development version seems to avoid problems.  Please "Paste, drop, or click to add files", or give a downloadable URL, of a current actual literal executable file which fails, so that the modified code can be tested.

**Comment by JMLX42** [comment]
> UPX has tried a few methods, and its current development version seems to avoid problems. Please "Paste, drop, or click to add files", or give a downloadable URL, of a current actual literal executable file which fails, so that the modified code can be tested.

@jreiser just so we're on the same page: you need "current actual literal executable" Rust executable linked with `mold` that fails when running `upx`?

**Comment by murlakatamenka** [comment]
@jreiser probably github doesn't allow attaching binaries, but making un-`upx`-able one is trivial:

```sh
cargo new hi

RUSTFLAGS='-C link-arg=-fuse-ld=mold' cargo build --release

# checking mold was used for linking
readelf -p .comment ./target/release/hi

upx --lzma --best ./target/releas/hi
```

Tested on Arch with Rust 1.82, mold v1.34:

```sh
rustc -V
# rustc 1.82.0 (f6e511eec 2024-10-15)

mold -v
# mold 2.34.1 (de65f6bd75b93eeb9f3775fc136ee79683ff35f1; compatible with GNU ld)
```

**Comment by murlakatamenka** [comment]
@jreiser probably github doesn't allow attaching binaries, but making un-`upx`-able one is trivial:

```sh
cargo new hi

RUSTFLAGS='-C link-arg=-fuse-ld=mold' cargo build --release

# checking mold was used for linking
readelf -p .comment ./target/release/hi

upx --lzma --best ./target/release/hi
```

Tested on Arch:

```sh
rustc -V
# rustc 1.82.0 (f6e511eec 2024-10-15)

mold -v
# mold 2.34.1 (de65f6bd75b93eeb9f3775fc136ee79683ff35f1; compatible with GNU ld)

upx -V
# upx 4.2.4
# UCL data compression library 1.03
# zlib data compression library 1.3.1.1-motley
# LZMA SDK version 4.43
```

**Comment by jreiser** [comment]
@murlakatamenka : GitHub does accept binary uploads in `.zip`, `.gz`, or `.tgz` format; so apply `zip`, `gzip`, or `tar -cz` to your executable before attaching.

Thank you for specifying versions of `rustc`, `mold`, `upx`.  That helps a lot, but I'm still faced with finding a "live USB" Arch (or the exact incantation to get an Arch environment under some virtual machine manager such as `docker`, `podman`, etc.), and the exact incantation to add the packages for `rustc` and `mold`.  You can help me get straight into debugging by uploading an actual file.  Uploading a specific actual file also facilitates community parallelism, by providing a reference for other users to compare against.

**Comment by murlakatamenka** [comment]
Yes, sure, here is the binary built as described [above](https://github.com/rust-lang/rust/issues/100859#issuecomment-2430737081):

```sh
readelf -p .comment ./target/release/hi
```

```
String dump of section '.comment':
  [     0]  GCC: (GNU) 14.2.1 20240910
  [    1c]  mold 2.34.1 (de65f6bd75b93eeb9f3775fc136ee79683ff35f1; compatible with GNU ld)
  [    6b]  rustc version 1.82.0 (f6e511eec 2024-10-15)
  [    97]  GCC: (GNU) 14.2.1 20240805
```

[hi.zip](https://github.com/user-attachments/files/17522723/hi.zip)
